Library/Homebrew/linkage_checker.rb

```diff
@@ -152,9 +155,18 @@ def check_dylibs(rebuild_cache:)
     checked_dylibs = Set.new
 
     keg_files_dylibs.each do |file, dylibs|
+      file_has_any_rpath_dylibs = T.let(false, T::Boolean)
       dylibs.each do |dylib|
         @reverse_links[dylib] << file
 
+        # Binary executables that link @rpath-prefixed dylibs must include at
+        # least one rpath in order to resolve it.
+        if !file_has_any_rpath_dylibs && (dylib.start_with? "@rpath/")
+          file_has_any_rpath_dylibs = true
+          pathname = Pathname(file)
+          @executables_missing_rpaths << file if pathname.binary_executable? && pathname.rpaths.empty?
```

Comment:
It might be good to not restrict this to executables, since this can be an issue for libraries too, which can turn into an issue for executables.

For example, executable `foo` links with `/path/to/libbar.dylib`, but `libbar.dylib` links with `@rpath/libbaz.dylib`. If `libbar` does not have an `LC_RPATH` entry for `libbaz` then `foo` is also broken.

Reply:
That's an interesting case I hadn't considered.

Opening this check up to all binaries found that `mpdecimal`'s `libmpdec++.2.5.1.dylib` references `@rpath/libmpdec.3.dylib` but does not specify an `LC_RPATH`. In this case, it would still be possible for the `LC_RPATH` of the executable or another dependent dylib to resolve this.

Perhaps it would be better to walk the tree of backrefs until we find an `LC_RPATH`?

Reply:
> In this case, it would still be possible for the LC_RPATH of the executable or another dependent dylib to resolve this.

Is it? I would find it surprising if Apple allows another binary's `LC_RPATH` entry to affect how the linker resolves library references. I tried testing it out locally too, and I don't think this is the case.

Reply:
My understanding is that the run-path search paths specified by executables are meant to resolve run-path dependent libraries loaded into the process.
https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/RunpathDependentLibraries.html

What isn't spelled out in the documentation is whether the relationship between run-path search path and run-path dependent library only applies for direct library references, or whether it applies to recursively referenced libraries as well. That said, the documentation makes no mention of a _library_ specifying a run-path search path for run-path dependent libraries it references.

Let me do some testing and see what works and what doesn't, at least as of macOS 12.

Reply:
When executable `a` links `@rpath\libb.dylib` which links `@rpath\libc.dylib`, a search path specified in either `a` or `libb.dylib` can be used to resolve `@rpath\libc.dylib`.

Relative search paths are evaluated relative to the path of the executable, even when specified in a dylib found in a different path. For example, if executable `.\a` links `@rpath\libb.dylib` specifying search path `foo`, and `foo\libb.dylib` links `@rpath\libc.dylib` specifying search path `bar`, `libc.dylib` must be in `foo\` or `bar` (relative `a`) to be resolved. It will not be resolved from `foo\bar` (`bar` relative to `foo\libb.dylib`).

Given that a dylib can legitimately reference run-path dependent libraries without specifying a search path, letting the executable's search path resolve the reference, I think that we can really only perform this check on executables.

Reply:
I'm not sure -- I don't think we want to ship libraries that contain `@rpath` references that rely on the user knowing to pass the right `-Wl,-rpath` flags to link the executables against the library. It would probably make linker failures easier to debug too.

Reply:
I agree that makes sense. I think it's clear enough that it's a Homebrew requirement and not an OS requirement. I've expanded the check to include executables, dylibs, and bundles.